paddle/fluid/operators/conv_op.mn.cc

```diff
@@ -0,0 +1,274 @@
+
+    const T* input_data = input->data<T>();
+    const T* filter_data = filter->data<T>();
+    // allocate memory for output
+    T* output_data = output->mutable_data<T>(ctx.GetPlace());
+
+    PADDLE_ENFORCE(input->dims().size() == 4,
+                   "Input must be with 4 dimensions, i.e. NCHW");
+    PADDLE_ENFORCE(filter->dims().size() == 4,
+                   "Filter must be with 4 dimensions, i.e. OIHW");
+
+    std::vector<int> src_tz = paddle::framework::vectorize2int(input->dims());
+    std::vector<int> weights_tz =
+        paddle::framework::vectorize2int(filter->dims());
+    std::vector<int> dst_tz = paddle::framework::vectorize2int(output->dims());
+
+    // MKLDNN primitives
+    // memory descriptors for convolution src/weight/dst
+    memory::dims conv_src_tz = {src_tz[0], src_tz[1], src_tz[2], src_tz[3]};
+    memory::dims conv_weights_tz = {weights_tz[0], weights_tz[1], weights_tz[2],
+                                    weights_tz[3]};
+    memory::dims conv_dst_tz = {dst_tz[0], dst_tz[1], dst_tz[2], dst_tz[3]};
+
+    memory::dims conv_strides = {strides[0], strides[1]};
+    memory::dims conv_padding = {paddings[0], paddings[1]};
+
+    auto conv_src_md = memory::desc({conv_src_tz}, memory::data_type::f32,
+                                    memory::format::nchw);
+    auto conv_weights_md = memory::desc(
+        {conv_weights_tz}, memory::data_type::f32, memory::format::oihw);
+    auto conv_dst_md = memory::desc({conv_dst_tz}, memory::data_type::f32,
+                                    memory::format::nchw);
+
+    // create memory primitives
+    auto conv_src_memory =
+        memory({conv_src_md, mkldnn_engine}, (void*)input_data);
+    auto conv_weights_memory =
+        memory({conv_weights_md, mkldnn_engine}, (void*)filter_data);
+    auto conv_dst_memory = memory({conv_dst_md, mkldnn_engine}, output_data);
+
+    // create convolution op descriptor
+    auto conv_desc = convolution_forward::desc(
+        prop_kind::forward, convolution_direct, conv_src_md, conv_weights_md,
+        conv_dst_md, conv_strides, conv_padding, conv_padding,
+        padding_kind::zero);
+
+    // conv primitive desc need be used in backward path
+    // so we need allocate it in heap (instead of in stack)
+    convolution_forward::primitive_desc* p_conv_pd;
+    p_conv_pd =
+        new convolution_forward::primitive_desc(conv_desc, mkldnn_engine);
+    // save conv_pd into dev_ctx to be referred in backward path
+    std::shared_ptr<void> conv_pd;
+    conv_pd.reset(p_conv_pd);
+    dev_ctx.SetBlob(key_conv_pd, conv_pd);
+
+    // create convolution op primitive
+    auto conv_prim = convolution_forward(*p_conv_pd, conv_src_memory,
+                                         conv_weights_memory, conv_dst_memory);
+
+    // push op to stream and wait MKLDNN until it's executed
+    std::vector<primitive> pipeline{conv_prim};
+    stream(stream::kind::eager).submit(pipeline).wait();
```

Comment:
Only `conv_pd` is saved to context, I think we can save more, like `engine`,  `primitives`, `stream`, etc.

Reply:
I'm going to refactor it as soon as we know how to handle data transferring between forward and backward in parallel mode (ParallelDo OP).

Reply:
refactoring is done